### PR TITLE
Use absolute verification method URI for doc data

### LINKED ISSUE #205: [Request] Generate absolute URI for `verificationMethod`
## Description
when generating a VC the `proof.verificationMethod` just has the fragment identifier i.e. `#key` but not the full DID URI pointing to the method, not sure if it should contain the full URI or what the standard says at that respect

## Motivation
Alignment with W3C standard

## Requirements
Write a list of what you want this feature to do.

1. Required for proofs
2.
3.

## Are you planning to do it yourself in a pull request?

No

### PR DESCRIPTION
**Pull request by DeppLearning**
- add TryMethod trait
- closes #205

# Description of change
VCs and presentations use absolute verification method URIs, while documents and diffchains stick to relative URIS per default. VCs that come from outside our library, i.e. via parsing jsons can still contain absolute URLs.

## Links to any relevant issues
fixes issue #205

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [X] I have followed the contribution guidelines for this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by l1h3r** [comment]
This would be much simpler if it just used `method.id()` everywhere and in the future could allow signing one document with another.

**Review by JelleMillenaar — APPROVED** [review]
Flags a conflict with the latest dev branch. Please remerge dev into the branch. After that is done and fixed, the branch can be merged. I think its a good solution for now.

File: identity-credential/src/credential/credential.rs [hunk 1]
```diff
@@ -194,6 +196,10 @@ impl<T> SetSignature for Credential<T> {
     self.proof.replace(value);
   }
 }
+
+impl<T> TryMethod for Credential<T> {
+  const TYPE: MethodUriType = MethodUriType::Absolute;
+}
```
Comment (l1h3r) [inline identity-credential/src/credential/credential.rs:202]: 💅 newline

File: identity-did/src/verifiable/tests.rs [hunk 2]
```diff
@@ -52,6 +54,9 @@ impl SetSignature for That {
     self.proof = Some(signature);
   }
 }
+impl TryMethod for That {
```
Comment (l1h3r) [inline identity-did/src/verifiable/tests.rs:57]: 💅 newline

File: identity-did/src/verification/method.rs [hunk 3]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
...
 }
+
+/// Represents all possible verification method types
+///
+/// see [W3C DID-core spec](https://www.w3.org/TR/did-core/#relative-did-urls)
+pub enum MethodUriType {
+  Absolute,
+  Relative,
+}
+/// Used to return absolute or relative method URI.
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:143]: 💅 newline

File: identity-did/src/verification/method.rs [hunk 4]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
...
+///
+/// This trait is used to determine whether absolute or relative method URIs
+/// should be used to signing data.
+///
+/// [More Info](https://www.w3.org/TR/did-core/#relative-did-urls)
+pub trait TryMethod {
+  /// Flag that determines whether absolute or rleative URI
+  const TYPE: MethodUriType;
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:151]: 💅 newline

File: identity-did/src/verification/method.rs [hunk 5]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
...
+  Absolute,
+  Relative,
+}
+/// Used to return absolute or relative method URI.
+///
+/// This trait is used to determine whether absolute or relative method URIs
+/// should be used to signing data.
+///
+/// [More Info](https://www.w3.org/TR/did-core/#relative-did-urls)
+pub trait TryMethod {
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:149]: This trait and relevant enum can probably go in their own file

File: identity-iota/src/did/doc/document.rs [hunk 6]
```diff
@@ -56,6 +58,9 @@ pub struct Document {
   message_id: MessageId,
 }
 
+impl TryMethod for Document {
+  const TYPE: MethodUriType = MethodUriType::Absolute;
+}
```
Comment (l1h3r) [inline identity-iota/src/did/doc/document.rs:63]: 💅 newline

File: identity-did/src/verification/method.rs [hunk 7]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
...
+  /// Flag that determines whether absolute or rleative URI
+  const TYPE: MethodUriType;
+  /// Returns String representation of absolute or relative method URI, if any.
+  fn method(method_id: &DID) -> Option<String> {
+    let fragment = method_id.fragment()?;
+
+    Some(match Self::TYPE {
+      MethodUriType::Absolute => method_id.to_string(),
+      MethodUriType::Relative => core::iter::once('#').chain(fragment.chars()).collect(),
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:158]: if you change the argument to `method: &Method<T>` you won't have to re-implement `try_into_fragment` 😉

File: identity-did/src/verification/method.rs [hunk 8]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
     self.id()
   }
 }
+
+/// Represents all possible verification method types
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:136]:
```suggestion
/// Represents all possible verification method URI types
```

File: identity-did/src/verification/method.rs [hunk 9]
```diff
@@ -132,3 +132,39 @@ impl<T> AsRef<DID> for Method<T> {
...
+///
+/// see [W3C DID-core spec](https://www.w3.org/TR/did-core/#relative-did-urls)
+pub enum MethodUriType {
+  Absolute,
+  Relative,
+}
+/// Used to return absolute or relative method URI.
+///
+/// This trait is used to determine whether absolute or relative method URIs
+/// should be used to signing data.
```
Comment (l1h3r) [inline identity-did/src/verification/method.rs:146]: signing -> sign

**Review by l1h3r — APPROVED** [review]
💯

**Pull request merged**